Added some documentation

DH-DoveG opened the pull request:
Added a list of documents:

1. LuaState
2. LuaTable
3. LuaFunction
4. LuaCoroutine
5. LuaError
6. LuaObject

In addition to the above classes, documentation files for other classes are initialized.
Sorry, since English is not my native language, there may be some mistakes or ambiguities in the document, if so, please let me know.

gilzoide commented on doc/LuaScript.xml:
```diff
@@ -0,0 +1,16 @@
+	<methods>
+		<method name="new" qualifiers="const vararg">
+			<return type="Variant" />
+			<description>
```
A description similar to [GDScript.new](https://docs.godotengine.org/en/stable/classes/class_gdscript.html#class-gdscript-method-new)'s would be fine here.

DH-DoveG replied:
Yes, this is a good idea! I missed it at first. Now I have supplemented it and made an example of using it.
```
# lua_script.lua:
# ```lua
# local obj = {}
# obj.attack_point = 20
# obj.life_point = 100
# function obj:add_life(value)
# self.life_point = self.life_point + value
# end
# return obj
# ```
var lua_script = load("lua_script.lua")
var instance = lua_script.new()
print(instance.attack_point) # Prints 20
instance.add_life(10)
print(instance.get("life_point")) # Prints 110
```

gilzoide commented on doc/LuaObject.xml:
```diff
@@ -0,0 +1,28 @@
+		Base class for all Lua objects in GDScript, except LuaState.
+	</brief_description>
+	<description>
+		The LuaObject class provides methods to access the associated Lua state and retrieve object pointers.
+		This class serves as the foundation for interacting with Lua objects from GDScript. It allows you to:
+		- Get the Lua state associated with the object
+		- Retrieve the pointer value of the Lua object
```
Not sure if it's really necessary to talk about the internal object's pointer value here 🤔 If you think it's relevant, feel free to keep it, though.

gilzoide commented on doc/LuaObject.xml:
```diff
@@ -0,0 +1,28 @@
+		<method name="get_lua_state" qualifiers="const">
+			<return type="LuaState" />
+			<description>
+				Returns the LuaState associated with this Lua object.
```
I think the LuaState being the "owner" and the LuaObject being "associated" to it makes more sense. It's just about the wording, in this case. Maybe something like "Returns the LuaState that owns this Lua object.", what do you think?

DH-DoveG replied:
Yes, I think what you said makes the relationship between them clearer.
> Returns the LuaState that owns this Lua object.

gilzoide commented on doc/LuaObject.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<class name="LuaObject" inherits="RefCounted" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://raw.githubusercontent.com/godotengine/godot/master/doc/class.xsd">
+	<brief_description>
+		Base class for all Lua objects in GDScript, except LuaState.
```
`LuaError` is also not a LuaObject. Not sure if it's relevant here, though, as people can find what classes inherit from LuaObject easily anyway 🤔

DH-DoveG replied:
Yes, now I have modified the description to make it more in line with the actual situation.  However, I haven't conducted a more in-depth study of this class, so it might still not be appropriate.

gilzoide replied:
Maybe something like "Base class for objects that exist in a Lua virtual machine"?

DH-DoveG replied:
Yes, it's a good idea!

gilzoide reviewed (commented):
Looks and works great overall, thank you very much for the contribution! I've added comments around what I think should be changed to match the addon's behavior, and some extra suggestions and/or discussions.

DH-DoveG commented on src/gen/doc_data.gen.cpp:
```diff
@@ -0,0 +1,10 @@
+/* THIS FILE IS GENERATED DO NOT EDIT */
```
Yes, I tried to delete it further. It was completely fine. We actually didn't need this file anymore, so I deleted it.

gilzoide replied:
> We actually didn't need this file anymore, so I deleted it.

Nice, thanks! I see you only deleted it, could you please also add it to `.gitignore` so that we never add it again by mistake?

DH-DoveG replied:
Yes, it's important to ignore the cpp files that are generated every time we compile. I've added them to the .gitignore file and renamed the generated folder to 'generated-document' to avoid confusion. I tried compiling locally, and yes, it worked—the files are now being ignored.

DH-DoveG commented on doc/LuaTable.xml:
```diff
@@ -0,0 +1,118 @@
+			<description>
+				Returns the number of elements in the table.
+			</description>
+		</method>
```
Thank you for your correction. It's indeed as you said. I revised the document and attached an example. I think the example is helpful.

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+			Can be used in GDScript.
+		</member>
+	</members>
+	<constants>
```
Thank you for your inspection. It has been supplemented now

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+			<description>
+				Loads Lua code from a file and returns it as a [LuaFunction].
+				[param filename] Supports both absolute and relative paths. Defaults to searching in the "res://" directory.
+				[param buffer_size] Determines the number of bytes read per file operation. Larger values use more memory but reduce read operations.
```
Thank you for your inspection. It has been modified now.

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+			<param index="2" name="env" type="LuaTable" default="null" />
+			<description>
+				Executes Lua code from a string.
+				Returns a [Variant] if the execution produces a result. Returns a [LuaError] if there are compilation errors.
```
Thank you for your inspection. It has been modified now.

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+				Executes Lua code from a file.
+				[param filename] Supports both absolute and relative paths. Defaults to searching in the "res://" directory.
+				[param buffer_size] Determines the number of bytes read per file operation. Larger values use more memory but reduce read operations.
+				Returns a [Variant] if the execution produces a result. Returns a [LuaError] if there are compilation errors.
```
Thank you for your inspection. It has been modified now.

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+		<method name="get_lua_exec_dir" qualifiers="static">
+			<return type="String" />
+			<description>
+				Returns the directory where the Lua executable is located.
```
Thank you for your inspection. It has been modified now.

DH-DoveG commented on doc/LuaState.xml:
```diff
@@ -0,0 +1,316 @@
+			<return type="LuaFunction" />
+			<param index="0" name="callable" type="Callable" />
+			<description>
+				Creates and returns a new [LuaFunction] from the given [Callable]. This method does not require any [enum Library] to be opened.
```
Yes, I think it's good to delete it because `create_table` doesn't need to open any libraries either. This might not need to be emphasized.

DH-DoveG commented:
Some documents have been modified.
The omissions were supplemented.

### List:

1. LuaState
2. LuaTable
3. LuaObject
4. LuaScript

The following are screenshots of the modified parts：

### LuaObject
![图片](https://github.com/user-attachments/assets/521dd8bc-8a4e-4115-b6e1-51336e99903d)

### LuaTable
![图片](https://github.com/user-attachments/assets/f9626065-1bfe-4e8f-901b-06cf60a52863)

### LuaScript
![图片](https://github.com/user-attachments/assets/f4949cfb-75eb-4a32-9350-08a40954bd2a)

### LuaState
![图片](https://github.com/user-attachments/assets/6635baea-d1f2-4ff6-86a6-0ed188571524)

gilzoide commented:
Thanks for the updates @DH-DoveG, looks great! The only thing I'll require you to do before I merge this is ignoring the generated C++ file, to make sure we don't add it by mistake do Git (which also avoids the annoyance of it showing up in Git clients as a new file). The comment on LuaObject is more of a suggestion, I'm not sure if it's better or not that what you proposed.

DH-DoveG commented:
Thanks. Yes, it's important to ignore the cpp files that are generated every time we compile. I've added them to the .gitignore file and renamed the generated folder to 'generated-document' to avoid confusion. I tried compiling locally, and yes, it worked—the files are now being ignored.

Honestly, I don't have any better ideas for LuaObject right now. How about we leave it as is until we come up with something better? What do you think @gilzoide ?"

### LuaFunction
I noticed that the documentation for the method to_callable of LuaFunction was missing while reviewing the documents. I have added it now.

### LuaCoroutine
 I added an example to LuaCoroutine. I hope this will help people use this class.

DH-DoveG commented:
Sorry, I just saw this.  Yes, it's a good idea, and it's more suitable than what we have now.  Let's use this.  `LuaObject` has already been modified accordingly.
> Maybe something like "Base class for objects that exist in a Lua virtual machine"?

gilzoide commented on doc/LuaCoroutine.xml:
```diff
@@ -6,6 +6,29 @@
+		        for i = 1, size do
+		            count = i + count
+		            count = coroutine.yield(count)
+		    end
+		end
```
This indentation doesn't look right.

DH-DoveG replied:
Thanks, the indentation issue has been fixed now.

gilzoide commented:
Thanks for the updates @DH-DoveG, looks great! Now there's only a minor indentation problem in the new LuaCoroutine example (which was a very nice addition, btw), but after that, I think it'll be ready to merge 🚀

DH-DoveG commented:
Yes, I think it's ready now!

gilzoide commented:
I think it is! I'll just let the CI build this and I'll merge afterwards.

Thanks again, documentation is looking great \o/
